src/ctx.rs

```diff
@@ -57,42 +81,146 @@ impl AppConfig {
             .unwrap_or(DEFAULT_TOKEN_EXPIRATION_TIME)
     }
 
-    pub(crate) fn get_proxy_route_by_inbound(&self, inbound: String) -> Option<&ProxyRoute> {
+    pub(crate) fn get_proxy_route_by_inbound(&self, inbound: &str) -> Option<&ProxyRoute> {
         let route_index = self.proxy_routes.iter().position(|r| {
-            r.inbound_route == inbound || r.inbound_route.to_owned() + "/" == inbound
+            r.inbound_route == inbound
+                || r.inbound_route == "/".to_owned() + inbound
+                || r.inbound_route.to_owned() + "/" == inbound
+                || "/".to_owned() + &*r.inbound_route == "/".to_owned() + inbound
         });
 
-        if let Some(index) = route_index {
-            return Some(&self.proxy_routes[index]);
+        route_index.map(|index| &self.proxy_routes[index])
+    }
+
+    /// Finds the best matching proxy route based on the provided URI's path and updates the URI by
+    /// removing the matched path segments while preserving the query parameters.
+    pub(crate) fn get_proxy_route_extracting_uri_inbound(
+        &self,
+        uri: &mut Uri,
+    ) -> GenericResult<Option<&ProxyRoute>> {
+        let path_segments: Vec<&str> = uri.path().split('/').filter(|s| !s.is_empty()).collect();
+
+        let mut best_match: Option<(&ProxyRoute, usize)> = None;
+
+        for r in &self.proxy_routes {
+            let route_segments: Vec<&str> = r
+                .inbound_route
+                .split('/')
+                .filter(|s| !s.is_empty())
+                .collect();
+            // Count the number of segments that match between the route and the path
+            let matched_segments = route_segments
+                .iter()
+                .zip(&path_segments)
+                .take_while(|(route_seg, path_seg)| route_seg == path_seg)
+                .count();
+            // Update best_match if this route fully matches (all its segments are matched)
+            // and best_match is None or has more matched segments than the current best match
+            if matched_segments == route_segments.len() // Ensure all segments of the route are matched
+                && (best_match.is_none() || matched_segments > best_match.unwrap().1)
+            {
+                best_match = Some((r, matched_segments));
+            }
+        }
+
+        if let Some((route, matched_segments)) = best_match {
+            // Construct the remaining path by skipping matched segments and accumulating the rest
+            let remaining_path: String = path_segments.iter().skip(matched_segments).fold(
+                String::new(),
+                |mut acc, segment| {
```

Comment:
You know the total size already - Can you consider preallocating `String` size here?